File: Models/SortFilterProxyModel.fs
module FSharpQt.Models.SortFilterProxyModel

open FSharpQt.BuilderNode
open System
open Org.Whatever.MinimalQtForFSharp
open FSharpQt.MiscTypes

open FSharpQt.Attrs

type Signal =
    | AutoAcceptChildRowsChanged of autoAcceptChildRows: bool
    | FilterCaseSensitivityChanged of filterCaseSensitivity: CaseSensitivity
    | FilterRoleChanged of filterRole: ItemDataRole
    | RecursiveFilteringEnabledChanged of recursiveFilteringEnabled: bool
    | SortCaseSensitivityChanged of sortCaseSensitivity: CaseSensitivity
    | SortLocaleAwareChanged of sortLocaleAware: bool
    | SortRoleChanged of sortRole: ItemDataRole
    
type internal Attr =
    | AutoAcceptChildRows of state: bool
    | DynamicSortFilter of state: bool
    | FilterCaseSensitivity of sensitivity: CaseSensitivity
    | FilterKeyColumn of column: int option
    | FilterRegularExpression of regex: Regex
    | FilterRole of role: ItemDataRole
    | SortLocaleAware of state: bool
    | RecursiveFilteringEnabled of state: bool
    | SortCaseSensitivity of sensitivity: CaseSensitivity
    | SortRole of role: ItemDataRole
with
    interface IAttr with
        override this.AttrEquals other =
            match other with
            | :? Attr as otherAttr ->
                this = otherAttr
            | _ ->
                false
        override this.Key =
            match this with
            | AutoAcceptChildRows _ -> "sortfilterproxymodel:autoacceptchildrows"
            | DynamicSortFilter _ -> "sortfilterproxymodel:dynamicsortfilter"
            | FilterCaseSensitivity _ -> "sortfilterproxymodel:filtercasesensitivity"
            | FilterKeyColumn _ -> "sortfilterproxymodel:filterkeycolumn"
            | FilterRegularExpression _ -> "sortfilterproxymodel:filterregularexpression"
            | FilterRole _ -> "sortfilterproxymodel:filterrole"
            | SortLocaleAware _ -> "sortfilterproxymodel:sortlocaleaware"
            | RecursiveFilteringEnabled _ -> "sortfilterproxymodel:recursivefilteringenabled"
            | SortCaseSensitivity _ -> "sortfilterproxymodel:sortcasesensitivity"
            | SortRole _ -> "sortfilterproxymodel:sortrole"
        override this.ApplyTo (target: IAttrTarget, maybePrev: IAttr option) =
            match target with
            | :? AttrTarget as attrTarget ->
                attrTarget.ApplySortFilterProxyModelAttr(this)
            | _ ->
                printfn "warning: SortFilterProxyModel.Attr couldn't ApplyTo() unknown target type [%A]" target
                
and internal AttrTarget =
    interface
        inherit AbstractProxyModel.AttrTarget
        abstract member ApplySortFilterProxyModelAttr: Attr -> unit
    end
  
type private SignalMapFunc<'msg>(func) =
    inherit SignalMapFuncBase<Signal,'msg>(func)

type Props<'msg>() =
    inherit AbstractProxyModel.Props<'msg>()
    
    let mutable onAutoAcceptChildRowsChanged: (bool -> 'msg) option = None
    let mutable onFilterCaseSensitivityChanged: (CaseSensitivity -> 'msg) option = None
    let mutable onFilterRoleChanged: (ItemDataRole -> 'msg) option = None
    let mutable onRecursiveFilteringEnabledChanged: (bool -> 'msg) option = None
    let mutable onSortCaseSensitivityChanged: (CaseSensitivity -> 'msg) option = None
    let mutable onSortLocaleAwareChanged: (bool -> 'msg) option = None
    let mutable onSortRoleChanged: (ItemDataRole -> 'msg) option = None
    
    member internal this.SignalMask = enum<SortFilterProxyModel.SignalMask> (int this._signalMask)
    
    member this.AutoAcceptChildRowsChanged with set value =
        onAutoAcceptChildRowsChanged <- Some value
        this.AddSignal(int SortFilterProxyModel.SignalMask.AutoAcceptChildRowsChanged)

    member this.FilterCaseSensitivityChanged with set value =
        onFilterCaseSensitivityChanged <- Some value
        this.AddSignal(int SortFilterProxyModel.SignalMask.FilterCaseSensitivityChanged)

    member this.FilterRoleChanged with set value =
        onFilterRoleChanged <- Some value
        this.AddSignal(int SortFilterProxyModel.SignalMask.FilterRoleChanged)

    member this.RecursiveFilteringEnabledChanged with set value =
        onRecursiveFilteringEnabledChanged <- Some value
        this.AddSignal(int SortFilterProxyModel.SignalMask.RecursiveFilteringEnabledChanged)

    member this.SortCaseSensitivityChanged with set value =
        onSortCaseSensitivityChanged <- Some value
        this.AddSignal(int SortFilterProxyModel.SignalMask.SortCaseSensitivityChanged)

    member this.SortLocaleAwareChanged with set value =
        onSortLocaleAwareChanged <- Some value
        this.AddSignal(int SortFilterProxyModel.SignalMask.SortLocaleAwareChanged)

    member this.SortRoleChanged with set value =
        onSortRoleChanged <- Some value
        this.AddSignal(int SortFilterProxyModel.SignalMask.SortRoleChanged)
        
    member internal this.SignalMapList =
        let thisFunc = function
            | AutoAcceptChildRowsChanged autoAcceptChildRows ->
                onAutoAcceptChildRowsChanged
                |> Option.map (fun f -> f autoAcceptChildRows)
            | FilterCaseSensitivityChanged filterCaseSensitivity ->
                onFilterCaseSensitivityChanged
                |> Option.map (fun f -> f filterCaseSensitivity)
            | FilterRoleChanged filterRole ->
                onFilterRoleChanged
                |> Option.map (fun f -> f filterRole)
            | RecursiveFilteringEnabledChanged recursiveFilteringEnabled ->
                onRecursiveFilteringEnabledChanged
                |> Option.map (fun f -> f recursiveFilteringEnabled)
            | SortCaseSensitivityChanged sortCaseSensitivity ->
                onSortCaseSensitivityChanged
                |> Option.map (fun f -> f sortCaseSensitivity)
            | SortLocaleAwareChanged sortLocaleAware ->
                onSortLocaleAwareChanged
                |> Option.map (fun f -> f sortLocaleAware)
            | SortRoleChanged sortRole ->
                onSortRoleChanged
                |> Option.map (fun f -> f sortRole)
        // prepend to parent signal map funcs
        SignalMapFunc(thisFunc) :> ISignalMapFunc :: base.SignalMapList
        
    member this.AutoAcceptChildRows with set value =
        this.PushAttr(AutoAcceptChildRows value)

    member this.DynamicSortFilter with set value =
        this.PushAttr(DynamicSortFilter value)

    member this.FilterCaseSensitivity with set value =
        this.PushAttr(FilterCaseSensitivity value)

    member this.FilterKeyColumn with set value =
        this.PushAttr(FilterKeyColumn value)

    member this.FilterRegularExpression with set value =
        this.PushAttr(FilterRegularExpression value)

    member this.FilterRole with set value =
        this.PushAttr(FilterRole value)

    member this.SortLocaleAware with set value =
        this.PushAttr(SortLocaleAware value)

    member this.RecursiveFilteringEnabled with set value =
        this.PushAttr(RecursiveFilteringEnabled value)

    member this.SortCaseSensitivity with set value =
        this.PushAttr(SortCaseSensitivity value)

    member this.SortRole with set value =
        this.PushAttr(SortRole value)
        
type ModelCore<'msg>(dispatch: 'msg -> unit) =
    inherit AbstractProxyModel.ModelCore<'msg>(dispatch)
    let mutable sfProxyModel: SortFilterProxyModel.Handle = null
    let mutable signalMap: Signal -> 'msg option = (fun _ -> None)
    let mutable currentMask = enum<SortFilterProxyModel.SignalMask> 0
    
    // binding guards
    let mutable lastAutoAcceptChildRows = false
    let mutable lastFilterCaseSensitivity = CaseSensitive
    let mutable lastFilterRole = DisplayRole
    let mutable lastRecursiveFilteringEnabled = false
    let mutable lastSortCaseSensitivity = CaseSensitive
    let mutable lastLocaleAware = false
    let mutable lastSortRole = DisplayRole
    
    let signalDispatch (s: Signal) =
        signalMap s
        |> Option.iter dispatch
        
    member this.SortFilterProxyModel
        with get() = sfProxyModel
        and set value =
            // assign to base
            this.AbstractProxyModel <- value
            sfProxyModel <- value
            
    member internal this.SignalMaps with set (mapFuncList: ISignalMapFunc list) =
        match mapFuncList with
        | h :: etc ->
            match h with
            | :? SignalMapFunc<'msg> as smf ->
                signalMap <- smf.Func
            | _ ->
                failwith "SortFilterProxyModel.ModelCore.SignalMaps: wrong func type"
            // assign the remainder to parent class(es)
            base.SignalMaps <- etc
        | _ ->
            failwith "SortFilterProxyModel.ModelCore: signal map assignment didn't have a head element"
            
    member this.SignalMask with set value =
        if value <> currentMask then
            // we don't need to invoke the base version, the most derived widget handles the full signal stack from all super classes (at the C++/C# levels)
            sfProxyModel.SetSignalMask(value)
            currentMask <- value
            
    interface AttrTarget with
        member this.ApplySortFilterProxyModelAttr attr =
            match attr with
            | AutoAcceptChildRows state ->
                if state <> lastAutoAcceptChildRows then
                    lastAutoAcceptChildRows <- state
                    sfProxyModel.SetAutoAcceptChildRows(state)
            | DynamicSortFilter state ->
                sfProxyModel.SetDynamicSortFilter(state)
            | FilterCaseSensitivity sensitivity ->
                if sensitivity <> lastFilterCaseSensitivity then
                    lastFilterCaseSensitivity <- sensitivity
                    sfProxyModel.SetFilterCaseSensitivity(sensitivity.QtValue)
            | FilterKeyColumn column ->
                sfProxyModel.SetFilterKeyColumn(column |> Option.defaultValue -1)
            | FilterRegularExpression regex ->
                sfProxyModel.SetFilterRegularExpression(regex.QtValue)
            | FilterRole role ->
                if role <> lastFilterRole then
                    lastFilterRole <- role
                    sfProxyModel.SetFilterRole(role.QtValue)
            | SortLocaleAware state ->
                if state <> lastLocaleAware then
                    lastLocaleAware <- state
                    sfProxyModel.SetSortLocaleAware(state)
            | RecursiveFilteringEnabled state ->
                if state <> lastRecursiveFilteringEnabled then
                    lastRecursiveFilteringEnabled <- state
                    sfProxyModel.SetRecursiveFilteringEnabled(state)
            | SortCaseSensitivity sensitivity ->
                if sensitivity <> lastSortCaseSensitivity then
                    lastSortCaseSensitivity <- sensitivity
                    sfProxyModel.SetSortCaseSensitivity(sensitivity.QtValue)
            | SortRole role ->
                if role <> lastSortRole then
                    lastSortRole <- role
                    sfProxyModel.SetSortRole(role.QtValue)
                    
    interface SortFilterProxyModel.SignalHandler with
        // Object =========================
        member this.Destroyed(obj: Object.Handle) =
            (this :> Object.SignalHandler).Destroyed(obj)
        member this.ObjectNameChanged(name: string) =
            (this :> Object.SignalHandler).ObjectNameChanged(name)
        // AbstractItemModel ==============
        member this.ColumnsAboutToBeInserted (parent, first, last) =
            (this :> AbstractItemModel.SignalHandler).ColumnsAboutToBeInserted(parent, first, last)
        member this.ColumnsAboutToBeMoved (srcParent, srcStart, srcEnd, destParent, destColumn) =
            (this :> AbstractItemModel.SignalHandler).ColumnsAboutToBeMoved(srcParent, srcStart, srcEnd, destParent, destColumn)
        member this.ColumnsAboutToBeRemoved (parent, first, last) =
            (this :> AbstractItemModel.SignalHandler).ColumnsAboutToBeRemoved(parent, first, last)
        member this.ColumnsInserted (parent, first, last) =
            (this :> AbstractItemModel.SignalHandler).ColumnsInserted(parent, first, last)
        member this.ColumnsMoved (srcParent, srcStart, srcEnd, destParent, destColumn) =
            (this :> AbstractItemModel.SignalHandler).ColumnsMoved(srcParent, srcStart, srcEnd, destParent, destColumn)
        member this.ColumnsRemoved (parent, first, last) =
            (this :> AbstractItemModel.SignalHandler).ColumnsRemoved(parent, first, last)
        member this.DataChanged (topLeft, bottomRight, roles) =
            (this :> AbstractItemModel.SignalHandler).DataChanged(topLeft, bottomRight, roles)
        member this.HeaderDataChanged (orientation, first, last) =
            (this :> AbstractItemModel.SignalHandler).HeaderDataChanged(orientation, first, last)
        member this.LayoutAboutToBeChanged (parents, hint) =
            (this :> AbstractItemModel.SignalHandler).LayoutAboutToBeChanged(parents, hint)
        member this.LayoutChanged (parents, hint) =
            (this :> AbstractItemModel.SignalHandler).LayoutChanged(parents, hint)
        member this.ModelAboutToBeReset() =
            (this :> AbstractItemModel.SignalHandler).ModelAboutToBeReset()
        member this.ModelReset() =
            (this :> AbstractItemModel.SignalHandler).ModelReset()
        member this.RowsAboutToBeInserted (parent, start, ``end``) =
            (this :> AbstractItemModel.SignalHandler).RowsAboutToBeInserted(parent, start, ``end``)
        member this.RowsAboutToBeMoved (srcParent, srcStart, srcEnd, destParent, destRow) =
            (this :> AbstractItemModel.SignalHandler).RowsAboutToBeMoved(srcParent, srcStart, srcEnd, destParent, destRow)
        member this.RowsAboutToBeRemoved (parent, first, last) =
            (this :> AbstractItemModel.SignalHandler).RowsAboutToBeRemoved(parent, first, last)
        member this.RowsInserted (parent, first, last) =
            (this :> AbstractItemModel.SignalHandler).RowsInserted(parent, first, last)
        member this.RowsMoved (srcParent, srcStart, srcEnd, destParent, destRow) =
            (this :> AbstractItemModel.SignalHandler).RowsMoved(srcParent, srcStart, srcEnd, destParent, destRow)
        member this.RowsRemoved (parent, first, last) =
            (this :> AbstractItemModel.SignalHandler).RowsRemoved(parent, first, last)
        // AbstractProxyModel
        member this.SourceModelChanged() =
            (this :> AbstractProxyModel.SignalHandler).SourceModelChanged()
        // SortFilterProxyModel
        member this.AutoAcceptChildRowsChanged autoAccept =
            signalDispatch (AutoAcceptChildRowsChanged autoAccept)
        member this.FilterCaseSensitivityChanged sensitivity =
            signalDispatch (CaseSensitivity.From sensitivity |> FilterCaseSensitivityChanged)
        member this.FilterRoleChanged role =
            signalDispatch (ItemDataRole.From role |> FilterRoleChanged)
        member this.RecursiveFilteringEnabledChanged enabled =
            signalDispatch (RecursiveFilteringEnabledChanged enabled)
        member this.SortCaseSensitivityChanged sensitivity =
            signalDispatch (CaseSensitivity.From sensitivity |> SortCaseSensitivityChanged)
        member this.SortLocaleAwareChanged state =
            signalDispatch (SortLocaleAwareChanged state)
        member this.SortRoleChanged role =
            signalDispatch (ItemDataRole.From role |> SortRoleChanged)

    interface IDisposable with
        member this.Dispose() =
            sfProxyModel.Dispose()
            
type private Model<'msg>(dispatch: 'msg -> unit) as this =
    inherit ModelCore<'msg>(dispatch)
    let sfProxyModel = SortFilterProxyModel.Create(this)
    do
        this.SortFilterProxyModel <- sfProxyModel
    
    member this.AddSourceModel (model: AbstractItemModel.Handle) =
        sfProxyModel.SetSourceModel(model)
        
    member this.RemoveSourceModel () =
        // well if it gets deleted (as a dependency), won't that automagically remove it from us?
        ()

let private create (attrs: IAttr list) (signalMaps: ISignalMapFunc list) (dispatch: 'msg -> unit) (signalMask: SortFilterProxyModel.SignalMask) =
    let model = new Model<'msg>(dispatch)
    model.ApplyAttrs (attrs |> List.map (fun attr -> None, attr))
    model.SignalMaps <- signalMaps
    model.SignalMask <- signalMask
    model

let private migrate (model: Model<'msg>) (attrs: (IAttr option * IAttr) list) (signalMaps: ISignalMapFunc list) (signalMask: SortFilterProxyModel.SignalMask) =
    model.ApplyAttrs attrs
    model.SignalMaps <- signalMaps
    model.SignalMask <- signalMask
    model

let private dispose (model: Model<'msg>) =
    (model :> IDisposable).Dispose()
    
type SortFilterProxyModelBinding internal(handle: SortFilterProxyModel.Handle) =
    inherit AbstractProxyModel.AbstractProxyModelBinding(handle)
    
let bindNode (name: string) (map: Map<string, IViewBinding>) =
    match map.TryFind name with
    | Some (:? SortFilterProxyModelBinding as sortFilterProxyModel) ->
        sortFilterProxyModel
    | _ ->
        failwith "SortFilterProxyModel.bindNode fail"

type SortFilterProxyModel<'msg>() =
    inherit Props<'msg>()
    [<DefaultValue>] val mutable private model: Model<'msg>
    
    let mutable maybeSourceModel: IModelNode<'msg> option = None
    member this.SourceModel with set value = maybeSourceModel <- Some value
    
    member this.MigrateDeps (changeMap: Map<DepsKey, DepsChange>) =
        match changeMap.TryFind (StrKey "source") with
        | Some change ->
            match change with
            | Unchanged ->
                ()
            | Added ->
                this.model.AddSourceModel(maybeSourceModel.Value.QtModel)
            | Removed ->
                this.model.RemoveSourceModel()
            | Swapped ->
                this.model.RemoveSourceModel()
                this.model.AddSourceModel(maybeSourceModel.Value.QtModel)
        | None ->
            // neither side had one
            ()
    
    interface IModelNode<'msg> with
        override this.Dependencies =
            maybeSourceModel
            |> Option.map (fun node -> StrKey "source", node :> IBuilderNode<'msg>)
            |> Option.toList

        override this.Create dispatch buildContext =
            this.model <- create this.Attrs this.SignalMapList dispatch this.SignalMask
            
        override this.AttachDeps () =
            maybeSourceModel
            |> Option.iter (fun node ->
                this.model.AddSourceModel(node.QtModel))

        override this.MigrateFrom (left: IBuilderNode<'msg>) (depsChanges: (DepsKey * DepsChange) list) =
            let left' = (left :?> SortFilterProxyModel<'msg>)
            let nextAttrs = diffAttrs left'.Attrs this.Attrs |> createdOrChanged
            this.model <- migrate left'.model nextAttrs this.SignalMapList this.SignalMask
            this.MigrateDeps(depsChanges |> Map.ofList)

        override this.Dispose() =
            (this.model :> IDisposable).Dispose()

        override this.QtModel =
            this.model.SortFilterProxyModel
            
        override this.ContentKey =
            this.model.SortFilterProxyModel
            
        override this.Attachments =
            this.Attachments
            
        override this.Binding =
            this.MaybeBoundName
            |> Option.map (fun name ->
                name, SortFilterProxyModelBinding(this.model.SortFilterProxyModel))

       

File: MinimalQtForFSharp/server/cpp/MinimalQtForFSharpImpl/SortFilterProxyModel.cpp
#include "generated/SortFilterProxyModel.h"

#include <QSortFilterProxyModel>
#include <utility>
#include "util/SignalStuff.h"
#include "util/Macros.h"

#include "RegularExpressionInternal.h"

#define THIS ((SortFilterProxyModelWithHandler*)_this)

namespace SortFilterProxyModel
{
    class SortFilterProxyModelWithHandler : public QSortFilterProxyModel {
        Q_OBJECT
    private:
        std::shared_ptr<SignalHandler> handler;
        SignalMask lastMask = 0;
        std::vector<SignalMapItem<SignalMaskFlags>> signalMap = {
            // Object:
            { SignalMaskFlags::Destroyed, SIGNAL(destroyed(QObject)), SLOT(onDestroyed(QObject)) },
            { SignalMaskFlags::ObjectNameChanged, SIGNAL(objectNameChanged(QString)), SLOT(onObjectNameChanged(QString)) },
            // AbstractItemModel:
            { SignalMaskFlags::ColumnsAboutToBeInserted, SIGNAL(columnsAboutToBeInserted(QModelIndex,int,int)), SLOT(onColumnsAboutToBeInserted(QModelIndex,int,int)) },
            { SignalMaskFlags::ColumnsAboutToBeMoved, SIGNAL(columnsAboutToBeMoved(QModelIndex,int,QModelIndex,int)), SLOT(onColumnsAboutToBeMoved(QModelIndex,int,QModelIndex,int)) },
            { SignalMaskFlags::ColumnsAboutToBeRemoved, SIGNAL(columnsAboutToBeRemoved(QModelIndex,int,int)), SLOT(onColumnsAboutToBeRemoved(QModelIndex,int,int)) },
            { SignalMaskFlags::ColumnsInserted, SIGNAL(columnsInserted(QModelIndex,int,int)), SLOT(onColumnsInserted(QModelIndex,int,int)) },
            { SignalMaskFlags::ColumnsMoved, SIGNAL(columnsMoved(QModelIndex,int,int,QModelIndex,int)), SLOT(onColumnsMoved(QModelIndex,int,int,QModelIndex,int)) },
            { SignalMaskFlags::ColumnsRemoved, SIGNAL(columnsRemoved(QModelIndex,int,int)), SLOT(onColumnsRemoved(QModelIndex,int,int)) },
            { SignalMaskFlags::DataChanged, SIGNAL(dataChanged(QModelIndex,QModelIndex,QList<int>)), SLOT(onDataChanged(QModelIndex,QModelIndex,QList<int>)) },
            { SignalMaskFlags::HeaderDataChanged, SIGNAL(headerDataChanged(Orientation,int,int)), SLOT(onHeaderDataChanged(Orientation,int,int)) },
            { SignalMaskFlags::LayoutAboutToBeChanged, SIGNAL(layoutAboutToBeChanged(QList<QPersistentModelIndex>,LayoutChangeHint)), SLOT(onLayoutAboutToBeChanged(QList<QPersistentModelIndex>,LayoutChangeHint)) },
            { SignalMaskFlags::LayoutChanged, SIGNAL(layoutChanged(QList<QPersistentModelIndex>,LayoutChangeHint)), SLOT(onLayoutChanged(QList<QPersistentModelIndex>,LayoutChangeHint)) },
            { SignalMaskFlags::ModelAboutToBeReset, SIGNAL(modelAboutToBeReset()), SLOT(onModelAboutToBeReset()) },
            { SignalMaskFlags::ModelReset, SIGNAL(modelReset()), SLOT(onModelReset()) },
            { SignalMaskFlags::RowsAboutToBeInserted, SIGNAL(rowsAboutToBeInserted(QModelIndex,int,int)), SLOT(onRowsAboutToBeInserted(QModelIndex,int,int)) },
            { SignalMaskFlags::RowsAboutToBeMoved, SIGNAL(rowsAboutToBeMoved(ModelIndex,int,int,ModelIndex,int)), SLOT(onRowsAboutToBeMoved(ModelIndex,int,int,ModelIndex,int)) },
            { SignalMaskFlags::RowsAboutToBeRemoved, SIGNAL(rowsAboutToBeRemoved(QModelIndex,int,int)), SLOT(onRowsAboutToBeRemoved(QModelIndex,int,int)) },
            { SignalMaskFlags::RowsInserted, SIGNAL(rowsInserted(QModelIndex,int,int)), SLOT(onRowsInserted(QModelIndex,int,int)) },
            { SignalMaskFlags::RowsMoved, SIGNAL(rowsMoved(QModelIndex,int,int,QModelIndex,int)), SLOT(onRowsMoved(QModelIndex,int,int,QModelIndex,int)) },
            { SignalMaskFlags::RowsRemoved, SIGNAL(rowsRemoved(QModelIndex,int,int)), SLOT(onRowsRemoved(QModelIndex,int,int)) },
            // AbstractProxyModel:
            { SignalMaskFlags::SourceModelChanged, SIGNAL(sourceModelChanged()), SLOT(onSourceModelChanged()) },
            // SortFilterProxyModel:
            { SignalMaskFlags::AutoAcceptChildRowsChanged, SIGNAL(autoAcceptChildRowsChanged(bool)), SLOT(onAutoAcceptChildRowsChanged(bool)) },
            { SignalMaskFlags::FilterCaseSensitivityChanged, SIGNAL(filterCaseSensitivityChanged(CaseSensitivity)), SLOT(onFilterCaseSensitivityChanged(CaseSensitivity)) },
            { SignalMaskFlags::FilterRoleChanged, SIGNAL(filterRoleChanged(int)), SLOT(onFilterRoleChanged(int)) },
            { SignalMaskFlags::RecursiveFilteringEnabledChanged, SIGNAL(recursiveFilteringEnabledChanged(bool)), SLOT(onRecursiveFilteringEnabledChanged(bool)) },
            { SignalMaskFlags::SortCaseSensitivityChanged, SIGNAL(sortCaseSensitivityChanged(CaseSensitivity)), SLOT(onSortCaseSensitivityChanged(CaseSensitivity)) },
            { SignalMaskFlags::SortLocaleAwareChanged, SIGNAL(sortLocaleAwareChanged(bool)), SLOT(onSortLocaleAwareChanged(bool)) },
            { SignalMaskFlags::SortRoleChanged, SIGNAL(sortRoleChanged(int)), SLOT(onSortRoleChanged(int)) },
        };
    public:
        explicit SortFilterProxyModelWithHandler(std::shared_ptr<SignalHandler> handler) : handler(std::move(handler)) {}
        void setSignalMask(SignalMask newMask) {
            if (newMask != lastMask) {
                processChanges(lastMask, newMask, signalMap, this);
                lastMask = newMask;
            }
        }
    public slots:
        // Object =================
        void onDestroyed(QObject *obj) {
            handler->destroyed((Object::HandleRef)obj);
        }
        void onObjectNameChanged(const QString& name) {
            handler->objectNameChanged(name.toStdString());
        }
        // AbstractItemModel ======
        void onColumnsAboutToBeInserted(const QModelIndex& parent, int first, int last) {
            handler->columnsAboutToBeInserted(MODELINDEX(parent), first, last);
        };
        void onColumnsAboutToBeMoved(const QModelIndex& sourceParent, int sourceStart, int sourceEnd, const QModelIndex& destinationParent, int destinationColumn) {
            handler->columnsAboutToBeMoved(MODELINDEX(sourceParent), sourceStart, sourceEnd, MODELINDEX(destinationParent), destinationColumn);
        };
        void onColumnsAboutToBeRemoved(const QModelIndex& parent, int first, int last) {
            handler->columnsAboutToBeRemoved(MODELINDEX(parent), first, last);
        };
        void onColumnsInserted(const QModelIndex& parent, int first, int last) {
            handler->columnsInserted(MODELINDEX(parent), first, last);
        };
        void onColumnsMoved(const QModelIndex& sourceParent, int sourceStart, int sourceEnd, const QModelIndex& destinationParent, int destinationColumn) {
            handler->columnsMoved(MODELINDEX(sourceParent), sourceStart, sourceEnd, MODELINDEX(destinationParent), destinationColumn);
        };
        void onColumnsRemoved(const QModelIndex& parent, int first, int last) {
            handler->columnsRemoved(MODELINDEX(parent), first, last);
        };
        void onDataChanged(const QModelIndex &topLeft, const QModelIndex &bottomRight, const QList<int> &roles = QList<int>()) {
            std::vector<ItemDataRole> roles2;
            for (auto &role : roles) {
                roles2.push_back((ItemDataRole)role);
            }
            handler->dataChanged(MODELINDEX(topLeft), MODELINDEX(bottomRight), roles2);
        }
        void onHeaderDataChanged(Qt::Orientation orientation, int first, int last) {
            handler->headerDataChanged((Enums::Orientation)orientation, first, last);
        };
        void onLayoutAboutToBeChanged(const QList<QPersistentModelIndex>& parents, QAbstractItemModel::LayoutChangeHint hint) {
            std::vector<PersistentModelIndex::HandleRef> parents2;
            for (auto &parent : parents) {
                parents2.push_back(PMODELINDEX(parent));
            }
            handler->layoutAboutToBeChanged(parents2, (AbstractItemModel::LayoutChangeHint)hint);
        };
        void onLayoutChanged(const QList<QPersistentModelIndex>& parents, QAbstractItemModel::LayoutChangeHint hint) {
            std::vector<PersistentModelIndex::HandleRef> parents2;
            for (auto &parent : parents) {
                parents2.push_back(PMODELINDEX(parent));
            }
            handler->layoutChanged(parents2, (AbstractItemModel::LayoutChangeHint)hint);
        };
        void onModelAboutToBeReset() {
            handler->modelAboutToBeReset();
        };
        void onModelReset() {
            handler->modelReset();
        };
        void onRowsAboutToBeInserted(const QModelIndex& parent, int start, int end) {
            handler->rowsAboutToBeInserted(MODELINDEX(parent), start, end);
        };
        void onRowsAboutToBeMoved(const QModelIndex& sourceParent, int sourceStart, int sourceEnd, const QModelIndex& destinationParent, int destinationRow) {
            handler->rowsAboutToBeMoved(MODELINDEX(sourceParent), sourceStart, sourceEnd, MODELINDEX(destinationParent), destinationRow);
        };
        void onRowsAboutToBeRemoved(const QModelIndex& parent, int first, int last) {
            handler->rowsAboutToBeRemoved(MODELINDEX(parent), first, last);
        };
        void onRowsInserted(const QModelIndex& parent, int first, int last) {
            handler->rowsInserted(MODELINDEX(parent), first, last);
        };
        void onRowsMoved(const QModelIndex& sourceParent, int sourceStart, int sourceEnd, const QModelIndex& destinationParent, int destinationRow) {
            handler->rowsMoved(MODELINDEX(sourceParent), sourceStart, sourceEnd, MODELINDEX(destinationParent), destinationRow);
        };
        void onRowsRemoved(const QModelIndex& parent, int first, int last) {
            handler->rowsRemoved(MODELINDEX(parent), first, last);
        };
        // AbstractProxyModel =====
        void onSourceModelChanged() {
            handler->sourceModelChanged();
        };
        // SortFilterProxyModel ===
        void onAutoAcceptChildRowsChanged(bool autoAcceptChildRows) {
            handler->autoAcceptChildRowsChanged(autoAcceptChildRows);
        };
        void onFilterCaseSensitivityChanged(Qt::CaseSensitivity filterCaseSensitivity) {
            handler->filterCaseSensitivityChanged((Enums::CaseSensitivity)filterCaseSensitivity);
        };
        void onFilterRoleChanged(int filterRole) {
            handler->filterRoleChanged((ItemDataRole)filterRole);
        };
        void onRecursiveFilteringEnabledChanged(bool recursiveFilteringEnabled) {
            handler->recursiveFilteringEnabledChanged(recursiveFilteringEnabled);
        };
        void onSortCaseSensitivityChanged(Qt::CaseSensitivity sortCaseSensitivity) {
            handler->sortCaseSensitivityChanged((Enums::CaseSensitivity)sortCaseSensitivity);
        };
        void onSortLocaleAwareChanged(bool sortLocaleAware) {
            handler->sortLocaleAwareChanged(sortLocaleAware);
        };
        void onSortRoleChanged(int sortRole) {
            handler->sortRoleChanged((ItemDataRole)sortRole);
        };
    };

    void Handle_setAutoAcceptChildRows(HandleRef _this, bool state) {
        THIS->setAutoAcceptChildRows(state);
    }

    void Handle_setDynamicSortFilter(HandleRef _this, bool state) {
        THIS->setDynamicSortFilter(state);
    }

    void Handle_setFilterCaseSensitivity(HandleRef _this, CaseSensitivity sensitivity) {
        THIS->setFilterCaseSensitivity((Qt::CaseSensitivity)sensitivity);
    }

    void Handle_setFilterKeyColumn(HandleRef _this, int32_t filterKeyColumn) {
        THIS->setFilterKeyColumn(filterKeyColumn);
    }

    void Handle_setFilterRegularExpression(HandleRef _this, std::shared_ptr<RegularExpression::Deferred::Base> regex) {
        THIS->setFilterRegularExpression(RegularExpression::fromDeferred(regex));
    }

    void Handle_setFilterRole(HandleRef _this, ItemDataRole filterRole) {
        THIS->setFilterRole((int)filterRole);
    }

    void Handle_setSortLocaleAware(HandleRef _this, bool state) {
        THIS->setSortLocaleAware(state);
    }

    void Handle_setRecursiveFilteringEnabled(HandleRef _this, bool enabled) {
        THIS->setRecursiveFilteringEnabled(enabled);
    }

    void Handle_setSortCaseSensitivity(HandleRef _this, CaseSensitivity sensitivity) {
        THIS->setSortCaseSensitivity((Qt::CaseSensitivity)sensitivity);
    }

    void Handle_setSortRole(HandleRef _this, ItemDataRole sortRole) {
        THIS->setSortRole((int)sortRole);
    }

    void Handle_setSignalMask(HandleRef _this, SignalMask mask) {
        THIS->setSignalMask(mask);
    }

    void Handle_dispose(HandleRef _this) {
        delete THIS;
    }

    HandleRef create(std::shared_ptr<SignalHandler> handler) {
        return (HandleRef) new SortFilterProxyModelWithHandler(std::move(handler));
    }

    // subclass stuff ==================================================

#define SUBTHIS ((Subclassed*)_this)

    class Subclassed : public SortFilterProxyModelWithHandler {
    private:
        MethodMask methodMask;
        std::shared_ptr<MethodDelegate> methodDelegate;
    public:
        Subclassed(const std::shared_ptr<SignalHandler> &handler, const std::shared_ptr<MethodDelegate> &methodDelegate, MethodMask methodMask)
            : SortFilterProxyModelWithHandler(handler), methodMask(methodMask), methodDelegate(methodDelegate) {}
    protected:
        [[nodiscard]] bool filterAcceptsRow(int source_row, const QModelIndex &source_parent) const override {
            if (methodMask & MethodMaskFlags::FilterAcceptsRow) {
                return methodDelegate->filterAcceptsRow(source_row, MODELINDEX(source_parent));
            } else {
                return QSortFilterProxyModel::filterAcceptsRow(source_row, source_parent);
            }
        }
        [[nodiscard]] bool filterAcceptsColumn(int source_column, const QModelIndex &source_parent) const override {
            if (methodMask & MethodMaskFlags::FilterAcceptsColumn) {
                return methodDelegate->filterAcceptsColumn(source_column,MODELINDEX(source_parent));
            } else {
                return QSortFilterProxyModel::filterAcceptsColumn(source_column, source_parent);
            }
        }
        [[nodiscard]] bool lessThan(const QModelIndex &source_left, const QModelIndex &source_right) const override {
            if (methodMask & MethodMaskFlags::LessThan) {
                return methodDelegate->lessThan(MODELINDEX(source_left), MODELINDEX(source_right));
            } else {
                return QSortFilterProxyModel::lessThan(source_left, source_right);
            }
        }
    public:
        friend void Interior_invalidateColumnsFilter(InteriorRef _this);
        friend void Interior_invalidateRowsFilter(InteriorRef _this);
        friend void Interior_invalidateFilter(InteriorRef _this);
    };

    void Interior_invalidateColumnsFilter(InteriorRef _this) {
        SUBTHIS->invalidateColumnsFilter();
    }

    void Interior_invalidateRowsFilter(InteriorRef _this) {
        SUBTHIS->invalidateRowsFilter();
    }

    void Interior_invalidateFilter(InteriorRef _this) {
        SUBTHIS->invalidateFilter();
    }

    void Interior_dispose(InteriorRef _this) {
        delete THIS;
    }

    InteriorRef createSubclassed(std::shared_ptr<SignalHandler> handler, std::shared_ptr<MethodDelegate> methodDelegate, MethodMask methodMask) {
        return (InteriorRef) new Subclassed(handler, methodDelegate, methodMask);
    }

}

#include "SortFilterProxyModel.moc"
